### PR TITLE
Introduce authentication in Legend DB Server

### PR DESCRIPTION
**Pull request by shtirlets**
#### What type of PR is this?
Improvement

#### What does this PR do / why is it needed ?
Kerberos identity
Push identity to Legend

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/handler/legend/LegendPreparedStatement.java [hunk 1]
```diff
@@ -44,7 +50,20 @@ public void setObject(int i, Object o) throws Exception
...
     public PostgresResultSetMetaData getMetaData() throws Exception
     {
-        return new LegendResultSetMetaData(client.getSchema(query));
+        if (identity.getFirstCredential() instanceof LegendKerberosCredential)
+
+        {
+            LegendKerberosCredential credential = (LegendKerberosCredential) identity.getFirstCredential();
+            return Subject.doAs(credential.getSubject(), (PrivilegedAction<LegendResultSetMetaData>) () ->
+            {
+                return new LegendResultSetMetaData(client.getSchema(query));
```
Comment (gs-manvig) [inline legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/handler/legend/LegendPreparedStatement.java:59]: Code in the `doAs` block and the else block look like duplicates. Can we extract that lambda to a private function and reuse it?

File: legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/handler/legend/LegendPreparedStatement.java [hunk 2]
```diff
@@ -68,10 +87,25 @@ public void setMaxRows(int maxRows) throws Exception
...
-        columns = schemaAndResult.getOne();
-        tdsRows = schemaAndResult.getTwo();
-        return true;
+        if (identity.getFirstCredential() instanceof LegendKerberosCredential)
+
+        {
+            LegendKerberosCredential credential = (LegendKerberosCredential) identity.getFirstCredential();
+            return Subject.doAs(credential.getSubject(), (PrivilegedAction<Boolean>) () ->
+            {
+                Pair<List<LegendColumn>, Iterable<TDSRow>> schemaAndResult = client.getSchemaAndExecuteQuery(query);
```
Comment (gs-manvig) [inline legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/handler/legend/LegendPreparedStatement.java:96]: Same comment as above

File: legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/handler/legend/LegendStatement.java [hunk 3]
```diff
@@ -17,28 +17,49 @@
...
-        Pair<List<LegendColumn>, Iterable<TDSRow>> schemaAndResult = client.getSchemaAndExecuteQuery(query);
-        columns = schemaAndResult.getOne();
-        tdsRows = schemaAndResult.getTwo();
-        return true;
+        if (identity.getFirstCredential() instanceof LegendKerberosCredential)
+        {
+            LegendKerberosCredential credential = (LegendKerberosCredential) identity.getFirstCredential();
+            return Subject.doAs(credential.getSubject(), (PrivilegedAction<Boolean>) () ->
+            {
+                Pair<List<LegendColumn>, Iterable<TDSRow>> schemaAndResult = client.getSchemaAndExecuteQuery(query);
```
Comment (gs-manvig) [inline legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/handler/legend/LegendStatement.java:49]: Same comment regarding duplicate code

File: legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/PostgresServer.java [hunk 4]
```diff
@@ -132,12 +121,41 @@ public NioEventLoopGroup getWorkerGroup()
     public static void main(String[] args)
     {
         int port = 9998;
+        String protocol;
```
Comment (gs-manvig) [inline legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/PostgresServer.java:124]: Given we are starting to have more configs, can we add config management for the db server?

**Review by gs-manvig — CHANGES_REQUESTED** [review]

File: legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/handler/legend/LegendTdsClient.java [hunk 5]
```diff
@@ -91,14 +93,19 @@ public Pair<List<LegendColumn>, Iterable<TDSRow>> getSchemaAndExecuteQuery(Strin
     private JsonNode executeQueryApi(String query)
     {
         System.out.println("executing query " + query);
-        try (CloseableHttpClient client = (CloseableHttpClient) HttpClientBuilder.getHttpClient(this.cookieStore))
+        try (CloseableHttpClient client = (CloseableHttpClient) HttpClientBuilder.getHttpClient(cookieStore))
         {
-            HttpPost req = new HttpPost("http://" + this.host + ":" + this.port + "/api/sql/v1/execution/execute/" + this.projectId);
+            String uri = protocol + "://" + this.host + ":" + this.port + "/api/sql/v1/execution/executeQueryString/" + this.projectId;
+            System.out.println(uri);
```
[the remark this answers is not in the document]
Comment (shtirlets) [reply legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/handler/legend/LegendTdsClient.java:99]: done

File: legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/PostgresServer.java (hunk 4 above)
Comment (shtirlets) [reply legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/PostgresServer.java:124]: done

File: legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/handler/legend/LegendStatement.java [hunk 6]
```diff
@@ -14,26 +14,48 @@
...
 
     @Override
     public boolean execute(String query) throws Exception
+    {
+        if (identity.getFirstCredential() instanceof LegendKerberosCredential)
+        {
+            LegendKerberosCredential credential = (LegendKerberosCredential) identity.getFirstCredential();
+            return Subject.doAs(credential.getSubject(), (PrivilegedAction<Boolean>) () ->
+            {
+                return esecute(query);
```
Comment (gs-manvig) [inline legend-engine-xt-sql-postgres-server/src/main/java/org/finos/legend/engine/postgres/handler/legend/LegendStatement.java:48]: Looks like a typo for execute

**Review by aormerod-gs — APPROVED** [review]

**Pull request merged**